### ISSUE TITLE
Quantized model is not working properly when CUBLAS is ON

### ISSUE THREAD

**Issue by bobqianic** [ggerganov/whisper.cpp#1661]
### Discussed in https://github.com/ggerganov/whisper.cpp/discussions/1656

<div type='discussions-op-text'>

<sup>Originally posted by **Sing303** December 19, 2023</sup>
Now when I try to use quantization models with FULL GPU CUBLAS, instead of recognizing text it writes nonsense, all kinds of signs instead of words.</div>

**Comment by Sing303** [comment]
At the same time we created :)
https://github.com/ggerganov/whisper.cpp/issues/1662

**Comment by bobqianic** [comment]
**Tried syncing the latest ggml from llama.cpp, but no luck. Both talk-llama and whisper are behaving identically.**

`tinyllama-1.1b-chat-v0.3.Q4_0.gguf` `ggml-model-whisper-base.bin`

![image](https://github.com/ggerganov/whisper.cpp/assets/129547291/280ebb16-2cd8-4d73-b41f-ab50b0b08a64)

`ggml-model-whisper-base-q5_1.bin`

![image](https://github.com/ggerganov/whisper.cpp/assets/129547291/83d71b75-040a-47ea-8961-b4935a8c960b)

**Comment by bobqianic** [comment]
The output logits of quantized models are near zero which is not correct. I dumped the `embd_enc` values, and I noticed that when using CUDA as the backend, the results significantly differ from those obtained using a CPU backend.

**Comment by bobqianic** [comment]
Do you know how to fix this bug? @slaren @ggerganov

**Comment by bobqianic** [comment]
I find it strange, if the error occurred during the `build_graph` process, then all backends should have a problem. Currently, it is known that only the CUDA backend has an issue. I have already synchronized the latest `ggml` from llama.cpp. But if the CUDA backend really has a problem, llama.cpp should have discovered this issue much earlier. Maybe it's because few people use it? I will try other quantization modes besides Q5_1.

**Comment by bobqianic** [comment]
Okay, I've tested all the quantization modes, and none of them work properly on CUDA. I wonder if our quantization method has changed? All `wstate_embd_enc` outputs have one thing in common: **`there is a significant difference`** between the outputs on CUDA and CPU backends, which does not exist in the properly functioning FP16 mode.

**Comment by slaren** [comment]
As far as I can tell, quantized models work correctly both with the gpt-2 example in ggml and in llama.cpp. I am not sure what is different in whisper.cpp.

**Comment by bobqianic** [comment]
`FP16`
![image](https://github.com/ggerganov/whisper.cpp/assets/129547291/0eb93378-cdae-4054-af44-2bc16c7386ae)

`Q8_0`
![image](https://github.com/ggerganov/whisper.cpp/assets/129547291/5074bb42-c8f6-4eca-9608-613905523641)

`Q6_K`
![image](https://github.com/ggerganov/whisper.cpp/assets/129547291/216a4c09-2184-42cb-bbf6-7eb2c98e3fd9)

`Q5_K`
![image](https://github.com/ggerganov/whisper.cpp/assets/129547291/62de4c16-2a40-453c-816f-74c355d25ca0)

`Q5_1`
![image](https://github.com/ggerganov/whisper.cpp/assets/129547291/db3be17d-7bb4-49cf-a67e-89855b49353d)

`Q5_0`
![image](https://github.com/ggerganov/whisper.cpp/assets/129547291/ba56f899-bdb1-4578-bcaa-0d37f01390e3)

`Q4_K`
![image](https://github.com/ggerganov/whisper.cpp/assets/129547291/5c703780-a115-4309-a539-ee3fad6e91b9)

`Q4_1`
![image](https://github.com/ggerganov/whisper.cpp/assets/129547291/9438d620-4b94-4e67-9022-44204f63465e)

`Q4_0`
![image](https://github.com/ggerganov/whisper.cpp/assets/129547291/07cd9dea-2351-4b83-aae2-5cd2f8002ed4)

`Q3_K`
![image](https://github.com/ggerganov/whisper.cpp/assets/129547291/44399eb2-10eb-43b4-b1cd-7501736ec3b4)

`Q2_K`
![image](https://github.com/ggerganov/whisper.cpp/assets/129547291/4a2e8995-f52a-4599-92b5-1478b66ff027)

**Comment by slaren** [comment]
Can you test changing the `GGML_CUDA_SOURCES` section in `CMakeLists.txt` to this (from ggml)? It may be an issue with the CUDA architectures.

```cmake
if (GGML_CUDA_SOURCES)
    message(STATUS "GGML CUDA sources found")
    if (NOT DEFINED CMAKE_CUDA_ARCHITECTURES)
        # Only configure gmml CUDA architectures is not globally set
        if (NOT DEFINED GGML_CUDA_ARCHITECTURES)
            # Not overriden by user, so set defaults
            set(GGML_CUDA_ARCHITECTURES 52 61 70)
        endif()
        message(STATUS "GGML Configuring CUDA architectures ${GGML_CUDA_ARCHITECTURES}")
        set_property(TARGET ggml  PROPERTY CUDA_ARCHITECTURES ${GGML_CUDA_ARCHITECTURES})
    endif()
    set_property(TARGET ggml  PROPERTY CUDA_SELECT_NVCC_ARCH_FLAGS "Auto")
    if (NOT MSVC)
        target_link_libraries(ggml PUBLIC stdc++)
    endif()
endif()
```

**Comment by bobqianic** [comment]
Is this the section you're referring to?

https://github.com/ggerganov/whisper.cpp/blob/9286d3f584240ba58bd44a1bd1e85141579c78d4/CMakeLists.txt#L522-L526

**Comment by slaren** [comment]
Yes. Replace the first `set_property` with this fragment:
```cmake
        # Only configure gmml CUDA architectures is not globally set
        if (NOT DEFINED GGML_CUDA_ARCHITECTURES)
            # Not overriden by user, so set defaults
            set(GGML_CUDA_ARCHITECTURES 52 61 70)
        endif()
        message(STATUS "GGML Configuring CUDA architectures ${GGML_CUDA_ARCHITECTURES}")
        set_property(TARGET ggml  PROPERTY CUDA_ARCHITECTURES ${GGML_CUDA_ARCHITECTURES})
```

**Comment by bobqianic** [comment]
Wow, it works!
